sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc

```diff
@@ -0,0 +1,650 @@
+  cgh.parallel_for<class bar>(range<1> { 1024 }, [=](id<1> idx) {
+    int data = my_pipe::read();
+    ...
+  });
+});
+
+my_pipe::write(5);  // OK.  Only communicates with single kernel
+----
+
+==== Illegal (host and multiple kernels connected with pipe):
+[source,c++,Host pipe restriction,linenums]
+----
+using my_pipe = pipe<class some_pipe, int>;
+
+myQueue.submit([&](handler& cgh) {
+  cgh.parallel_for<class foo>(range<1> { 1024 }, [=](id<1> idx) {
+    int data = my_pipe::read();
+    ...
+  });
+});
+
+myQueue.submit([&](handler& cgh) {
+  cgh.parallel_for<class bar>(range<1> { 1024 }, [=](id<1> idx) {
+    int data = my_pipe::read();
+    ...
+  });
+});
+
+my_pipe::write(5);  // Error.  Pipe communicates with two kernels
+----
+
+==== Illegal (host access combined with bidirectional access by a kernel):
+[source,c++,Host pipe restriction2,linenums]
+----
+using my_pipe = pipe<class some_pipe, int>;
+
+myQueue.submit([&](handler& cgh) {
+  cgh.parallel_for<class foo>(range<1> { 1024 }, [=](id<1> idx) {
+    int data = my_pipe::read();
+    my_pipe::write( data );  // Error: Write as well as read from kernel, on pipe that is also accessed from host code (below)
+  });
+});
+
+my_pipe::write(5);
+----
+
+==== Illegal (host both reads and writes pipe):
+[source,c++,Host pipe restriction3,linenums]
+----
+using my_pipe = pipe<class some_pipe, int>;
+
+my_pipe::write(5);
+int data = my_pipe::read();  // Error: Loopback pipes not allowed on host.  Pipes from host perspective must be unidirectional
+----
+
+
+== I/O Pipes
+
+The pipe class described above exposes both read and write static member functions.  Two additional classes are defined which can be exposed by an implementation, to provide access to hardware peripherals.  The link to a hardware peripheral is unidirectional, so the the classes that describe these links expose either read or write members, but not both.
+
+[source,c++,iopipes,linenums]
+----
+template <class name,
+          class dataT,
+          size_t min_capacity = 0>
+class kernel_readable_io_pipe {
+  public:
+    static dataT read();  // Blocking
+    static dataT read( bool &success_code );  // Non-blocking
+};
+
+template <class name,
+          class dataT,
+          size_t min_capacity = 0>
+class kernel_writeable_io_pipe {
+  public:
+    static void write( dataT data );  // Blocking
+    static void write( dataT data, bool &success_code );  // Non-blocking
+}
+----
+
+
+[source,c++,boardspec,linenums]
+----
+// "Built-in pipes" provide interfaces with hardware peripherals
+// These definitions are typically provided by a device vendor and
+// made available to developers for use.
+namespace example_platform {
+  template <unsigned ID>
+  struct ethernet_pipe_id {
+    static constexpr unsigned id = ID;
+  };
+  
+  using ethernet_read_pipe = kernel_readable_io_pipe<ethernet_pipe_id<0>, int, 0>;
+  using ethernet_write_pipe =kernel_writeable_io_pipe<ethernet_pipe_id<1>, int, 0>;
+}
+----
```

Comment:
> Just use some normal pipes with either a _read_ or _write_ attribute, but with some magical behaviour when they are named with an object coming from `sycl::vendor::...::ethernet` and so on.
I would prefer having separated classes. Otherwise, potential handling this in some backend would most likely looking like some ugly heuristics (IMHO).

Reply:
This is definitely a design question.
I do not know what you mean by "ugly heuristics". :-)
You just need to check that your pipe follows the PipeIO concept.
While waiting for C++20 :-) , you just need a type trait testing that your type has some specific implementation or whatever.
I am not sure that there is something so inherently different from a user perspective to have a separate type space.
Just like in Unix everything is a file in `/proc`, `/dev`  or `/sys`... Sure some files are more special and magical than others and have some `ioctl()` super power or whatever... :-)
But this is not a fundamental deal breaker for this proposal anyway.

Reply:
We started off this way, but switched to the wrapper style to simplify the interface and make less error prone